### PR TITLE
[ASA-340][SM]: UpdateAmls during subscription create and also Audit A…

### PR DESCRIPTION
**Pull request by sureshhmrc**
…mls details

### CODE REVIEW

File: conf/application.conf [hunk 1]
```diff
@@ -255,6 +265,11 @@ Prod {
         host = enrolment-store-proxy.protected.mdtp
         port = 443
       }
+      enrolment-store-proxy {
```
Comment (johnrevill) [inline conf/application.conf]: `agent-assurance` rather than `enrolment-store-proxy`

Comment (sureshhmrc) [reply conf/application.conf]: good catch - fixed it

File: app/uk/gov/hmrc/agentsubscription/service/SubscriptionService.scala [hunk 2]
```diff
@@ -84,41 +97,32 @@ class SubscriptionService @Inject() (
...
         if (agentRecord.isAnASAgent && postcodesMatch(agentRecord.businessPostcode, updateSubscriptionRequest.knownFacts.postcode)) {
           val subscriptionRequest = mergeSubscriptionRequest(updateSubscriptionRequest, agentRecord)
-
-          subscribe(subscriptionRequest, authIds, agentRecord.isAnASAgent, Some(agentRecord.arn))
+          for {
+            arn <- Some(agentRecord.arn) match {
+              case Some(arn) if agentRecord.isAnASAgent => Future.successful(arn)
+              case _ => desConnector.subscribeToAgentServices(subscriptionRequest.utr, desRequest(subscriptionRequest))
+            }
+            updatedAmlsDetails <- agentAssuranceConnector.updateAmls(updateSubscriptionRequest.utr, arn)
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/agentsubscription/service/SubscriptionService.scala]: the only difference between the above `createSubscription`  and this `updateSubscription` is  in the 
`createSubscription` we are creating amls record a s a last operation but here in the `updateSubscription` its here - and I cant move this operation further up as we need the `arn`  which apperas to be something wrong here ?

File: app/uk/gov/hmrc/agentsubscription/service/SubscriptionService.scala [hunk 3]
```diff
@@ -68,12 +68,25 @@ class SubscriptionService @Inject() (
...
-      }
+      case Some(DesRegistrationResponse(isAnAsAgent, _, _, maybeArn, BusinessAddress(_, _, _, _, Some(desPostcode), _), _)) if postcodesMatch(desPostcode, subscriptionRequest.knownFacts.postcode) =>
+        for {
+          arn <- maybeArn match {
+            case Some(arn) if isAnAsAgent => Future.successful(arn)
+            case _ => desConnector.subscribeToAgentServices(subscriptionRequest.utr, desRequest(subscriptionRequest))
+          }
+          _ <- addKnownFactsAndEnrol(arn, subscriptionRequest, authIds)
+          _ <- {
+            subscriptionRequest.amlsDetails match {
```
Comment (johnrevill) [inline app/uk/gov/hmrc/agentsubscription/service/SubscriptionService.scala]: I think these lines 81-84 need to be executed at the point just before the call to the desConnector.subscribeToAgentServices on line 77.

Comment (johnrevill) [reply app/uk/gov/hmrc/agentsubscription/service/SubscriptionService.scala]: And we also need to call `agentAssuranceConnector.updateAmls` after the call to `desConnector.subscribeToAgentServices` (and before the `addKnownFactsAndEnrol`)

Comment (sureshhmrc) [reply app/uk/gov/hmrc/agentsubscription/service/SubscriptionService.scala]: but `Arn` is not available before `desConnector.subscribeToAgentServices`

Comment (sureshhmrc) [reply app/uk/gov/hmrc/agentsubscription/service/SubscriptionService.scala]: ah..sorry.. hold on

File: app/uk/gov/hmrc/agentsubscription/connectors/AgentAssuranceConnector.scala [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+  def updateAmls(utr: Utr, arn: Arn)(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[Option[AmlsDetails]] = {
+
+    val url = new URL(baseUrl, s"/agent-assurance/amls/utr/${utr.value}")
+
+    monitor(s"ConsumedAPI-Put-AgentAssurance-Update-amls") {
+      http.PUT[JsObject, HttpResponse](url.toString, Json.obj("value" -> arn.value))
+        .map[Option[AmlsDetails]](r => Some(r.json.as[AmlsDetails]))
+        .recover {
+          case _: Upstream4xxResponse =>
```
Comment (johnrevill) [inline app/uk/gov/hmrc/agentsubscription/connectors/AgentAssuranceConnector.scala]: Small thing: Might need a condition on this match here to target the expected responses (e.g. 404, and 409?), like there is on the recover block for the `createAmls`.
i.e. we'd want other unexpected 4xx responses to result in a failed future.

File: it/uk/gov/hmrc/agentsubscription/controllers/SubscriptionAuditingSpec.scala [hunk 5]
```diff
@@ -30,17 +31,17 @@ class SubscriptionAuditingSpec extends BaseAuditSpec with Eventually with DesStu
...
-      eventually {
-        DataStreamStub.verifyAuditRequestSent(
-          AgentSubscriptionEvent.AgentSubscription,
-          expectedTags,
-          expectedDetails(utr))
-      }
+      DataStreamStub.verifyAuditRequestSent(
+        AgentSubscriptionEvent.AgentSubscription,
+        expectedTags,
+        expectedDetails(utr))
```
Comment (johnrevill) [inline it/uk/gov/hmrc/agentsubscription/controllers/SubscriptionAuditingSpec.scala]: Not quite sure what `expectedDetails` is.. Are we checking that the extra AMLS bit goes in to the audit event?

**Pull request merged**